### ISSUE TITLE
Returning empty values

### ISSUE THREAD

**Issue by slinkydeveloper** [cloudevents/sdk-java#123]
From https://github.com/cloudevents/sdk-java/issues/119 and https://github.com/cloudevents/sdk-java/pull/108#discussion_r411193730:

What technique should we use, around the sdk, to notify the user of an eventual empty return value?

1. `Optional` getters
2. Have an annotation called `@Nullable` which notifies the user of an eventual null return value
3. Just explicit in the Javadoc that a method can return null

**Comment by slinkydeveloper** [comment]
I'm fine both with the `@Nullable` annotation (as soon as it's internal to the sdk and not an external dependency) or just having the javadoc to explicitly mention that a method can return null.

**Comment by bsideup** [comment]
Yeah, I think JSR-305 `@io.cloudevents.lang.Nullable` meta-annotation (similar to [Spring's](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/Nullable.html), for example) would be the best option, so that the static analysis tools can help.

We should also mark the package as non-nullable by default

**Comment by slinkydeveloper** [comment]
If i understood correctly with JSR-305 we need to bring in an additional dependency, which I want to avoid as much as possible, even more for just an annotation.

**Comment by bsideup** [comment]
@slinkydeveloper the point of JSR-305 is that you don't need the dependency, but create your own annotation (and this meta-annotation is annotated with `provided` JSR-305 annotation)

**Comment by slinkydeveloper** [comment]
@bsideup the one that you linked from Spring uses `@Nonnull` from `javax.annotation`, can you provide an example/doc that explains how it works?

**Comment by bsideup** [comment]
The JVM spec allows you to keep the annotations without having them at runtime. it is used by JSR-305 to implement meta-annotations that are marked with `java.annotation` at compile time, but don't need them at runtime.

As you can see, Spring does not depend on it:
https://mvnrepository.com/artifact/org.springframework/spring-core/5.2.5.RELEASE

BTW it is also Kotlin-friendly.

**Comment by making** [comment]
Hi, here is an example that follows Spring's approach.
https://github.com/making/yavi/tree/develop/src/main/java/am/ik/yavi/jsr305
The dependecy on JSR-305 is optional.
https://github.com/making/yavi/blob/develop/pom.xml#L64-L68

**Comment by slinkydeveloper** [comment]
I was looking around for this jsr305 dep, it looks an old unmantained module without anybody actively mantaining it, it even looks like they didn't migrated the code from https://code.google.com/archive/p/jsr-305/source. I'm not sure it's really a good idea... Anybody else implemented this jsr and is actively mantaining it?

**Comment by bsideup** [comment]
One of the most commonly used is https://mvnrepository.com/artifact/com.google.code.findbugs/jsr305

Spring Framework and other top Java projects are using it:
https://github.com/spring-projects/spring-framework/blob/8265db0ae1d83a1e204e602fa96e542c2677977f/build.gradle#L348

Note that it is just a set of annotations.

**Comment by slinkydeveloper** [comment]
The fact that everybody uses it doesn't make it maintained :smile: Anyway, are you willing to provide a PR to add them to api package?

**Comment by bsideup** [comment]
@slinkydeveloper I would agree with you if not the usages tab of that dependency where you will find a ton of major Java projects ;)

> Anyway, are you willing to provide a PR to add them to api package?

Sure! Expect a PR soon

**Comment by bsideup** [comment]
Status update: not stale, will be working on the PR today.

**Comment by slinkydeveloper** [comment]
Now that #138 is merged, are you willing to apply the same changes to other modules that needs it (or at least, check if it's needed in other modules)?

**Comment by bsideup** [comment]
My main concern was the primary interface (`CloudEvent`), and I think it is fine to keep `Optional`s in implementations for better readability (unless someone complains about it).

That said, the same annotation can definitely be used in other modules 👍

**Comment by slinkydeveloper** [comment]
> That said, the same annotation can definitely be used in other modules +1

I would love to, to keep consistency across our modules

**Comment by slinkydeveloper** [comment]
@bsideup after #157 is in, are you willing to review all our usages of these annotations across all modules?
